WordPress/src/main/java/org/wordpress/android/widgets/WPNetworkImageView.java

```diff
@@ -195,42 +195,71 @@ private void loadImageIfNecessary(final boolean isInLayoutPass,
         // from the network.
         ImageLoader.ImageContainer newContainer = WordPress.imageLoader.get(mUrl,
-                new ImageLoader.ImageListener() {
-                    @Override
-                    public void onErrorResponse(VolleyError error) {
-                        showErrorImage();
-                        // keep track of URLs that 404 so we can skip them the next time
-                        int statusCode = VolleyUtils.statusCodeFromVolleyError(error);
-                        if (statusCode == 404) {
-                            mUrlSkipList.add(mUrl);
-                        }
-
-                        if (imageLoadListener != null) {
-                            imageLoadListener.onError();
-                        }
-                    }
+                new WPNetworkImageLoaderListener(mUrl, isInLayoutPass, imageLoadListener), 0, 0, getScaleType());
+        // update the ImageContainer to be the new bitmap container.
+        mImageContainer = newContainer;
+    }
+
+    /**
+     * Our implementation of ImageLoader.ImageListener that keeps a reference to the requested URL
+     * and makes sure we're setting the correct requested picture on response.
+     *
+     * Ref: https://github.com/wordpress-mobile/WordPress-Android/issues/5100
+     * This is a fix for those cases when WPNetworkImageView instances are used in UI items that can be recycled.
+     * The cell containing WPNetworkImageView could be recycled while the image request was still underway,
+     * so when the request completed it set the picture to the one requested prior to recycling.
+     */
+    private class WPNetworkImageLoaderListener implements ImageLoader.ImageListener {
+        private final String mRequestedURL;
+        private final ImageLoadListener mImageLoadListener;
+        private final boolean mIsInLayoutPass;
+
+        WPNetworkImageLoaderListener(final String requestedURL,
+                                     final boolean isInLayoutPass,
+                                     final ImageLoadListener imageLoadListener) {
+            mRequestedURL = requestedURL;
+            mIsInLayoutPass = isInLayoutPass;
+            mImageLoadListener = imageLoadListener;
+        }
 
+        @Override
+        public void onErrorResponse(VolleyError error) {
+            // keep track of URLs that 404 so we can skip them the next time
+            int statusCode = VolleyUtils.statusCodeFromVolleyError(error);
+            if (statusCode == 404) {
+                mUrlSkipList.add(mRequestedURL);
+            }
+
+            if (!mUrl.equals(mRequestedURL)) {
+                return;
+            }
+            showErrorImage();
+
+            if (mImageLoadListener != null) {
+                mImageLoadListener.onError();
+            }
+        }
+
+        @Override
+        public void onResponse(final ImageLoader.ImageContainer response, boolean isImmediate) {
+            if (!mUrl.equals(mRequestedURL)) {
```

Comment:
I had checked and didn't seem that could happen, since `mUrl` is only set in `setImageUrl`, and if you pass an empty value the request is not even made. 
Anyway, I've added those checks for future proofing the code . Thanks for checking this.

Ready for another pass.